src/kde/quick_index.hh

```diff
@@ -0,0 +1,145 @@
+              const double max_window_size, const size_t bins_per_dimension,
+              const bool domain_decomposed);
+
+  //! Collect Ghost Data
+  void collect_ghost_data(const std::vector<double> &local_data,
+                          std::vector<double> &local_ghost_data) const;
+
+  //! Override function of 3D array ghost data.
+  void collect_ghost_data(const std::vector<std::array<double, 3>> &local_data,
+                          std::vector<std::array<double, 3>> &local_ghost_data) const;
+
+  //! Override function for vector<vector<double> array ghost data.
+  void collect_ghost_data(const std::vector<std::vector<double>> &local_data,
+                          std::vector<std::vector<double>> &local_ghost_data) const;
+
+  //! Fetch list of coarse index values bound by the window
+  std::vector<size_t> window_coarse_index_list(const std::array<double, 3> &window_min,
+                                               const std::array<double, 3> &window_max) const;
+
+  //! Map local+ghost data to grid window
+  void map_data_to_grid_window(const std::vector<double> &local_data,
+                               const std::vector<double> &ghost_data,
+                               std::vector<double> &grid_data,
+                               const std::array<double, 3> &window_min,
+                               const std::array<double, 3> &window_max,
+                               const std::array<size_t, 3> &grid_bins, const std::string &map_type,
+                               const bool normalize, const bool bias) const;
+
+  //! Map local+ghost data to grid window for multi-dimensional data
+  void map_data_to_grid_window(const std::vector<std::vector<double>> &local_data,
+                               const std::vector<std::vector<double>> &ghost_data,
+                               std::vector<std::vector<double>> &grid_data,
+                               const std::array<double, 3> &window_min,
+                               const std::array<double, 3> &window_max,
+                               const std::array<size_t, 3> &grid_bins, const std::string &map_type,
+                               const bool normalize, const bool bias) const;
+
+  //! Map local+ghost data to grid window
+  void map_data_to_sphere_grid_window(
+      const std::vector<double> &local_data, const std::vector<double> &ghost_data,
+      std::vector<double> &grid_data, const std::array<double, 3> &sphere_center,
+      const std::array<double, 3> &window_min, const std::array<double, 3> &window_max,
+      const std::array<size_t, 3> &grid_bins, const std::string &map_type, const bool normalize,
+      const bool bias) const;
+
+  //! Map local+ghost data to grid window for multi-dimensional data
+  void map_data_to_sphere_grid_window(const std::vector<std::vector<double>> &local_data,
+                                      const std::vector<std::vector<double>> &ghost_data,
```

Comment:
While not a hard code style requirement, we often list all input arguments first and place mutable arguments at the end of the list.   Take it or leave it.